RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc

```diff
@@ -0,0 +1,285 @@
+  edm::InputTag OurSegmentsTag ("OurSegments");
+  edm::InputTag generalTracksTag ("generalTracks");
+  OurSegmentsToken_ = consumes<ME0SegmentCollection>(OurSegmentsTag);
+  generalTracksToken_ = consumes<reco::TrackCollection>(generalTracksTag);
+}
+
+ME0SegmentMatcher::~ME0SegmentMatcher() {}
+
+void ME0SegmentMatcher::produce(edm::Event& ev, const edm::EventSetup& setup) {
+
+    LogDebug("ME0SegmentMatcher") << "start producing segments for " << ++iev << "th event ";
+
+    //Getting the objects we'll need
+    using namespace edm;
+
+    ESHandle<MagneticField> bField;
+    setup.get<IdealMagneticFieldRecord>().get(bField);
+    ESHandle<Propagator> shProp;
+    setup.get<TrackingComponentsRecord>().get("SteppingHelixPropagatorAlong", shProp);
+
+    using namespace reco;
+
+    Handle<ME0SegmentCollection> OurSegments;
+    //ev.getByLabel("me0Segments","",OurSegments);
+    ev.getByToken(OurSegmentsToken_,OurSegments);
+
+    std::auto_ptr<std::vector<ME0Muon> > oc( new std::vector<ME0Muon> ); 
+    std::vector<ME0Muon> TempStore; 
+
+    Handle <TrackCollection > generalTracks;
+    //ev.getByLabel <TrackCollection> ("generalTracks", generalTracks);
+    ev.getByToken(generalTracksToken_,generalTracks);
+
+
+    int TrackNumber = 0;
+    
+    for (std::vector<Track>::const_iterator thisTrack = generalTracks->begin();
+	 thisTrack != generalTracks->end(); ++thisTrack,++TrackNumber){
+      //Initializing our plane
+
+      //Remove later
+      if (fabs(thisTrack->eta()) < 1.8) continue;
+
+      float zSign  = thisTrack->pz()/fabs(thisTrack->pz());
+
+      float zValue = 526.75 * zSign;
+
+      Plane *plane = new Plane(Surface::PositionType(0,0,zValue),Surface::RotationType());
+
+      //Getting the initial variables for propagation
+
+      int chargeReco = thisTrack->charge(); 
+      GlobalVector p3reco, r3reco;
+
+      p3reco = GlobalVector(thisTrack->outerPx(), thisTrack->outerPy(), thisTrack->outerPz());
+      r3reco = GlobalVector(thisTrack->outerX(), thisTrack->outerY(), thisTrack->outerZ());
+
+      AlgebraicSymMatrix66 covReco;
+      //This is to fill the cov matrix correctly
+      AlgebraicSymMatrix55 covReco_curv;
+      covReco_curv = thisTrack->outerStateCovariance();
+      FreeTrajectoryState initrecostate = getFTS(p3reco, r3reco, chargeReco, covReco_curv, &*bField);
+      getFromFTS(initrecostate, p3reco, r3reco, chargeReco, covReco);
+
+      //Now we propagate and get the propagated variables from the propagated state
+      SteppingHelixStateInfo startrecostate(initrecostate);
+      SteppingHelixStateInfo lastrecostate;
+
+      const SteppingHelixPropagator* ThisshProp = 
+	dynamic_cast<const SteppingHelixPropagator*>(&*shProp);
+	
+      //lastrecostate = ThisshProp->propagate(startrecostate, *plane);
+      //lastrecostate = ThisshProp->propagateWithPath(startrecostate, *plane);
+      ThisshProp->propagate(startrecostate, *plane,lastrecostate);
```

Comment:
I don't see a need to cast to SteppingHelixPropagator and use this non-standard interface.
Use Propagator base class methods instead.